### ISSUE TITLE
[Question] clarify iscsi situation

### ISSUE THREAD

**Issue by flokli** [longhorn/longhorn#1984]
For testing purporses, I tried deploying longhorn into a `kind` multi-node cluster.

longhorn started crashlooping, because `iscsi` isn't available.

I'm a bit confused - the docs only say:

> Longhorn does leverage iSCSI, so extra configuration of the node may be required. This may include the installation of open-iscsi or iscsiadm depending on the distribution.

But they don't go into detail where ISCSI is used. Is it also used for the communication between different replicas in "Figure 1. Read/write Data Flow between the Volume, Longhorn Engine, Replica Instances, and Disks", or Is it only used when exposing it over iSCSI via the iSCSI frontend?

If the latter applies, could it be made optional?

**Comment by yasker** [comment]
iSCSI is used to expose the block device presented by the Longhorn volume. So the iSCSI initiator is required on the node. 

iSCSI is not used in the communication between Longhorn components.

**Comment by flokli** [comment]
Not sure I understand. If I don't want to use the iSCSI frontend from https://longhorn.io/docs/0.8.0/install/iscsi/, do I still need the iscsi packages?

**Comment by yasker** [comment]
Yes, you need. It's required for using Longhorn with Kubernetes. See https://longhorn.io/docs/0.8.0/install/requirements/

**Comment by flokli** [comment]
That doesn't reply to my question :-) Does longhorn make use of any of the icsci packages if I don't use the iSCSI frontend?

**Comment by yasker** [comment]
As I said, yes, you need. The open-iscsi package is required if you want to use Longhorn with Kubernetes, even you don't want to use `iscsi` frontend. Since `blockdev` frontend depends on it.

**Comment by flokli** [comment]
Can you point me to the exact binaries that are invoked? I tried digging into `longhorn-manager`, but could only find some `mkfs` and `mount` invocations there, and in `vendor/k8s.io/kubernetes/pkg/util/mount/mount_linux.go`.

**Comment by yasker** [comment]
You can check https://github.com/longhorn/go-iscsi-helper/search?q=iscsiadm . And here is the code we use to start the frontend (expose the block device) https://github.com/longhorn/longhorn-engine/blob/9d21b822ecec363320d94091fd314bb36721eb42/pkg/frontend/tgt/frontend.go .

iSCSI related operation is a part of Longhorn Engine (who is the data plane), rather than Longhorn Manager (who is the management plane).

**Comment by flokli** [comment]
Okay, so reading through https://github.com/longhorn/go-iscsi-helper/blob/ee87992ec536c9471aebb6a107d52cd0490dcc1a/longhorndev/dev.go#L119, for both block device and iscsi frontend, the iscsi target is created (`d.scsiDevice.CreateTarget()`, and for block device, the initiator is started, too - `d.scsiDevice.StartInitiator()`).

Apparently, enging longhorn creates iscsi targets for all devices, not just those it exposes via iscsi itself. 

I still don't fully understand how the longhorn engines are started, and how the interaction with the host `isciadm` binary happens. I tried to look for pointers in https://raw.githubusercontent.com/longhorn/longhorn/master/deploy/longhorn.yaml, it seems like some of the hosts devices are mounted into the container at least, but I didn't grasp the full picture.

Anyhow, it explains why `iscsid` needs to run on the host at least - I opened https://github.com/longhorn/website/pull/228 to make this more obvious to people wondering like I initially did :-)

It'd be appreciated if the "Longhorn does leverage iSCSI" in https://longhorn.io/docs/0.8.0/concepts/#14-csi-plugin could also be a bit extended, to better describe how iscsi is used, and how the interaction from longhorn engine with the host looks like.

**Comment by yasker** [comment]
@flokli A little bit of the background story.

In the early year of Longhorn, we were trying to find a way to expose the block device to the end-user. Since Longhorn is running in userspace, and the block device must be exposed by the Linux kernel, we essentially have a few different ways of doing it:
1. NDB. Unreliable at the time. Can cause kernel crash easily.
2. FUSE + Loopback device. Very slow.
3. TCMU (TCM userspace). It's the latest in-kernel ISCSI target implementation, and TCMU provides a way to expose a userspace data plane as a block device. But, it was unstable, slow, and rarely used at the time (around 2016). We've made a few patches into the kernel to help increase its performance (included in Linux v4.6 IIRC), but since it was not widely used, we are afraid there are unknown bugs that might cause kernel crashes. Also, it's hard for the users to upgrade or patch the Linux kernel. Even though it's the fastest way we've found, we decided to move on.
4. Userspace iSCSI target + iSCSI initiator on the host. This is the current implementation. We've used a battle-tested userspace iSCSI target framework `tgt` to create an iSCSI endpoint for the Longhorn data plane, then we take advantage of the iSCSI initiator on the host (by switching namespace to the host namespace) to create a block device connecting to the iSCSI target exposed by `tgt` on the same node. Since the iSCSI initiator doesn't support cgroup/namespace in Linux, so we have to rely on the user to install it. Luckily it's installed by default for many Linux server distros. The performance is not as good as TCMU, but it's not bad at all, and we don't need to worry about iSCSI initiator cause kernel crash since it's definitely battle-tested.

In the end, we chose `tgt` + iSCSI initiator (`iscsid + iscsiadm`) on the node.

Also, in theory, you can say you only want Longhorn to expose the iSCSI target and never use the iSCSI initiator, and it might still work (a.k.a `iscsi` frontend only, no `blockdev` frontend). But:
1. Longhorn's Kubernetes driver is designed to work with `blockdev` frontend only
2. And even if you only want to use `iscsi` frontend, you still need the iSCSI initiator for the block device.

That's why we require iSCSI initiator to be installed on the nodes.

**Comment by flokli** [comment]
Thanks a lot for writing up the detailled explanation! I'd love if some parts of this could end up somewhere on the website, too :-)

**Comment by flokli** [comment]
@yasker is there something more to add to the website/docs from your explanation? I might not be the only one bumping into this, and curious about the design decisions.
